nomt/src/page_cache.rs

```diff
@@ -0,0 +1,251 @@
+                    .map(Arc::new)
+                    .map_or(Page::Nil, Page::Exists);
+                let mut shared = shared.lock().unwrap();
+                // Unwrap: the operation was inserted above. It is scheduled for execution only
+                // once. It may removed only in the line below. Therefore, `None` is impossible.
+                let inflight = shared.inflight.remove(&page_id).unwrap();
+                shared.prestine.insert(page_id, entry.clone());
+                inflight.complete_and_notify(entry);
+                drop(shared);
+            }
+        };
+        self.fetch_tp.execute(task);
+        promise
+    }
+
+    /// Flushes all the dirty pages into the underlying store.
+    ///
+    /// After the commit, all the dirty pages are cleared.
+    pub fn commit(&self, tx: &mut Transaction) {
+        let mut shared = self.shared.lock().unwrap();
+        for (page_id, page) in mem::take(&mut shared.dirty) {
+            shared.prestine.insert(page_id, page.clone());
+            let page_data = match page {
+                Page::Nil => None,
+                Page::Exists(data) => Some(data),
+            };
+            tx.write_page(page_id, page_data.as_deref());
+            // It doesn't seem that the page can be in the inflight fetches at this point.
+            //
+            // If the page ended up in the dirty set, then it must have been supplanted in case
+            // there was an inflight fetch for that page. After it was supplanted, the `retrieve`
+            // function would not initiate another fetch for the same page.
+            //
+            // Therefore, we don't need to divert the inflight fetches here.
+        }
+    }
+
+    /// Creates a new cursor assuming the given root node.
+    pub fn create_cursor(&self, root: Node) -> PageCacheCursor {
+        PageCacheCursor {
+            root,
+            page_cache: self.clone(),
+        }
+    }
+
+    /// Replaces the page in the cache with the given page.
+    ///
+    /// The inflight fetch for the given page if any is resolved with the given page.
+    pub fn supplant(&self, page_id: PageId, page: Page) {
+        let mut shared = self.shared.lock().unwrap();
+        if let Some(inflight) = shared.inflight.get_mut(&page_id) {
+            inflight.supplant_and_notify(page);
+        }
+    }
+}
+
+pub struct PageCacheCursor {
+    root: Node,
+    page_cache: PageCache,
+    // TODO: position
+}
+
+impl PageCacheCursor {
+    /// Moves the cursor to the given [`KeyPath`].
+    ///
+    /// Moving the cursor using this function would be more efficient than using the navigation
+    /// functions such as [`Self::down_left`] and [`Self::down_right`] due to leveraging warmup
+    /// hints.
+    ///
+    /// After returning of this function, the cursor is positioned either at the given key path or
+    /// at the closest key path that is on the way to the given key path.
+    pub fn seek(&mut self, dest: KeyPath) {
```

Comment:
this does not look like a normal `seek` function, which typically seeks to a position without loading what is before. The sketch looks more like a `fn prefetch`, but that is more of a caching concern than a navigation concern.